cdp_scrapers/legistar_utils.py

```diff
@@ -375,6 +380,167 @@ def get_legistar_events_for_timespan(
     return response
 
 
+def get_legistar_content_uris(
+    client: str, legistar_ev: Dict
+) -> Optional[List[ContentURIs]]:
+    """
+    Return URLs for videos and captions from a Legistar/Granicus-hosted video web page
+
+    Parameters
+    ----------
+    client: str
+        Which legistar client to target. Ex: "seattle"
+    legistar_ev: Dict
+        Data for one Legistar Event.
+
+    Returns
+    -------
+    content_uris: Optional[List[ContentURIs]]
+        None if web page HTML structure does not meet expected Legistar/Granicus format.
+        Empty list if the there is no video for this event.
+
+    Raises
+    ------
+    NotImplementedError
+        Means the content structure of the web page hosting session video has changed.
+        We need explicit review and update the scraping code.
+    """
+    global video_page_parser
+
+    # prefer video file path in legistar Event.EventVideoPath
+    if legistar_ev[LEGISTAR_SESSION_VIDEO_URI]:
+        return [
+            ContentURIs(
+                video_uri=str_simplified(legistar_ev[LEGISTAR_SESSION_VIDEO_URI]),
+                caption_uri=None,
+            )
+        ]
+    if not legistar_ev[LEGISTAR_EV_SITE_URL]:
+        return None
+
+    try:
+        # a td tag with a certain id pattern.
+        # this is usually something like
+        # https://somewhere.legistar.com/MeetingDetail.aspx...
+        # that is a summary-like page for a meeting
+        with urlopen(legistar_ev[LEGISTAR_EV_SITE_URL]) as resp:
+            soup = BeautifulSoup(resp.read(), "html.parser")
+
+    except (URLError, HTTPError) as e:
+        log.debug(f"{legistar_ev[LEGISTAR_EV_SITE_URL]}: {str(e)}")
+        return None
+
+    # this gets us the url for the web PAGE containing the video
+    # video link is provided in the window.open()command inside onclick event
+    # <a id="ctl00_ContentPlaceHolder1_hypVideo"
+    # data-event-id="75f1e143-6756-496f-911b-d3abe61d64a5"
+    # data-running-text="In&amp;nbsp;progress" class="videolink"
+    # onclick="window.open('Video.aspx?
+    # Mode=Granicus&amp;ID1=8844&amp;G=D64&amp;Mode2=Video','video');
+    # return false;"
+    # href="#" style="color:Blue;font-family:Tahoma;font-size:10pt;">Video</a>
+    extract_url = soup.find(
+        "a",
+        id=re.compile(r"ct\S*_ContentPlaceHolder\S*_hypVideo"),
+        class_="videolink",
+    )
+    if extract_url is None:
+        return None
+    # the <a> tag will not have this attribute if there is no video
+    if "onclick" not in extract_url.attrs:
+        return []
```

Comment:
Should this be `[]` or `None`?

Reply:
I like this comment. Made the function return a status code for hopefully allow caller to handle the returned information more appropriately.

```Python
class ContetUriScrapeResult(NamedTuple):
    class Status(enum.IntEnum):
        # Web page(s) are in unrecognized structure
        UnrecognizedPatternError = -1
        # Error in accessing some resource
        ResourceAccessError = -2
        # Video was not provided for the event
        ContentNotProvidedError = -3
        # Found URIs to video and optional caption
        Ok = 0

    status: Status
    uris: Optional[List[ContentURIs]] = None
```

Not entirely necessary but I like my `enums` haha.